[verifying-client] Verify consistency proofs

phlip9 opened the pull request:
## Overview

This PR includes everything needed to support verifying consistency proofs in the verifying client, from exposing a get_accumulator_consistency_proof API in json-rpc and DbReader, to refactoring TrustedState so it can keep track of its trusted accumulator summary, to modifying the verifying client itself so that it can query an initial accumulator and verify consistency proofs when ratcheting state proofs.

## Commits

### [storage] Add get_accumulator_consistency_proof to DbReader

This diff allows DB clients to query a specific accumulator consistency proof for a given range, rather than via get_state_proof(start_version), which will only give a consistency proof to the latest li.

One notable change is that the starting version is an `Option<Version>`, where `None` denotes starting the proof from "pre-genesis".  This API change allows DB clients to not only get a range like `get_accumulator_consistency_proof(Some(100), 200)`, but also first initialize their accumulator summaries via `get_accumulator_consistency_proof(None, 100)`.  They can do this because a consistency proof from pre-genesis to a verseion is exactly the frozen subtrees in an accumulator summary valid at that version.


### [json-rpc] Add get_accumulator_consistency_proof API

Expose the new get_accumulator_consistency_proof API to json-rpc clients, which allows clients to initialize their in-memory accumulator summaries and query consistency proofs to extend their in-memory accumulator summaries to new versions.

The main use-case for this stack is enabling verifying clients to initialize their in-memory accumulator summary when first starting up.  Once initialized, they can then actually verify the accumulator consistency proofs from get_state_proof, which we are not currently verifying at all.

The second use-case, which will come in a separate PR, is supporting historical get_metadata(version) for verifying clients. We can use this API to request `get_accumulator_consistency_proof(None, version)` and `get_accumulator_consistency_proof(version, None)` to get a proof for the historical accumulator root hash.


### [sdk] Add get_accumulator_consistency_proof API to non-verifying client


### [storage] Return consistency proof only for client-verifiable range

When a verifying client is syncing via get_state_proof requests, they can't verify the accumulator consistency proof when there are more epoch changes needed to sync before entering the epoch of that latest li. The reason is that the current behaviour always returns a consistency proof to the latest li version, regardless of whether or not the client will actually be able to reach the latest li after applying the epoch changes.

This diff fixes the above issue by limiting the consistency proof to the last client-verifiable version. If the state proof manages to get the client to the latest li, then the consistency proof is as before.  Otherwise, if the client will need more epoch changes, then the consistency proof is only to the last epoch change li version.


### [types] Add TransactionAccumulatorSummary

Adds a small wrapper type around `InMemoryAccumulator<TransactionInfoHasher>` with methods for more easily handling the transaction accumulator use-case.


### [types] Add MockTransactionAccumulator

Adds a small in-memory transaction accumulator for testing. The main use-case here is to hold all leaf hashes and allow arbitrary historical root hash and consistency proof queries against the accumulator.


### [types] Verify accumulator consistency proofs in trusted state

This diff modifies the `TrustedState` to support verifying accumulator consistency proofs. Now when verifying state proofs, we will correctly verify the consistency proof up to the last verifiable ledger info.

In order to verify a consistency proof, the trusted state needs to keep track of a verified transaction accumulator summary; ratcheting a state proof will also move the accumulator summary forward to the newly trusted state.

When verifying from a waypoint, we haven't yet built an accumulator summary. `verify_and_ratchet` now also takes an optional, untrusted `initial_accumulator`, which the trusted state will verify and use when it does not yet have an accumulator summary built up.

One consequence of verifying the consistency proofs is that we can no longer verify a state proof with a newer latest li but a partially trusted/verified prefix of epoch changes, since consistency proofs can only be verified from the exact accumulator summary at their starting version. This means users like the verifying client must verify_and_ratchet from the trusted state they had at the start of their request rather than reading the current trusted state when they verify the response.

This diff also adds support for verifying the initializing the accumulator and verifying consistency proofs to the verifying client.


### [sdk] Less strict concurrency model for clients

Currently, issuing concurrent requests with either verifying or non-verifying clients is not super useful since the ordering requirement is too strict. Each concurrent request races to place its latest observed state into the state store / trusted state store; the losers are considered "stale" and must be retried.

This strategy means we're almost less efficient than issuing requests serially, since we have to keep retrying the losers and (worst case) only one request wins each round.

The benefit of this strategy is that we maintain a consistent and strictly increasing global ordering of responses, with the important caveat that this is only true as of the time we compare-and-exchange the response state into the respective store and the response is the winner. Of course, the client may actually observe a different ordering due to asynchrony between placing the response into the store and actually observing the response in the client application code, rendering the benefit less useful than initially expected.

With this change, we relax the ordering requirements. When issuing multiple concurrent requests we now guarantee:

1. Each response batch is fulfilled and verified at a ledger version that is greater than or equal to the current trusted version _at the time we made the request batch_, though not necessarily the globally most recent trusted ledger version.

2. Requests made serially within a single thread of execution appear strictly ordered, i.e., they were fulfilled and verified at monotonically increasing ledger versions (`v1 <= v2 <= ...`).

Consequently, without any other effort, multiple concurrent requests may have responses that appear inconsistent or out-of-order. For example, naively making concurrent `get_account(..)` requests will (most likely) show accounts at different ledger versions; even further, the first response you receive may show a more recent ledger version than the second response.

To avoid these issues, clients should pin a batch of concurrent requests to the same ledger version in order to avoid an inconsistent ledger view.

libra-action commented:
**This PR may have modified JSON-RPC server code.**

Breaking changes policy:

1. Chaning JSON-RPC method signatures or removing/modifying fields
   in the response in /V1 will not be accepted.
2. Adding fields is ok.
3. Adding a new JSON-RPC method is ok.

Please ensure you have also done the following to update the docs:

1. Update json-rpc/API-CHANGELOG.md.
2. Add/update type/method documentation under json-rpc/docs.
<!-- metadata: {"tag":"jsonrpc-compat"} -->

msmouse commented on storage/diemdb/src/lib.rs:
```diff
@@ -920,6 +920,17 @@ impl DbReader for DiemDB {
         })
     }
+
+    fn get_accumulator_consistency_proof(
```
Beautiful. There's a typo "verseion" in the commit message though.

msmouse commented on types/src/proof/definition.rs:
```diff
@@ -265,6 +265,88 @@ where
+        consistency_proof: &AccumulatorConsistencyProof,
+        target_li: &LedgerInfo,
+    ) -> Result<Self> {
+        let num_new_txns = target_li.version() - self.0.version();
```
Probably use `checked_sub`, just to be cautious?

msmouse commented on crypto/crypto/src/hash.rs:
```diff
@@ -253,6 +253,17 @@ impl HashValue {
+    /// generating basic mock hash values.
+    ///
+    /// Ex: HashValue::from_u64(0x1234) => HashValue([0, .., 0, 0x12, 0x34])
+    pub fn from_u64(v: u64) -> Self {
```
Guard with `#[cfg(any(test, feature = "fuzzing"))]` would be nice.

kchalkias replied:
Yes this flag is very important to avoid accidental misuse of this public api. Note that the crypto lib, is available as a standalone crate in crates.io.

msmouse commented on types/src/proof/accumulator/mock.rs:
```diff
@@ -0,0 +1,174 @@
+/// Note: intended for test-only code; panics whenever something goes wrong.
+#[derive(Clone, Debug)]
+pub struct MockTransactionAccumulator {
+    nodes: BTreeMap<Position, HashValue>,
```
`HashMap` should suffice? And `Ord` and `PartialOrd` should not be needed.

----

Or if we really intend to utilize `Position::post_order_idx()`, a vector should do.

----
But I'd prefer not to use Position at all here, trying to be as different with production code as possible, and as straightforward as possible (examples: `NaiveSmt`, `NaiveSubtree` in the codebase.). Possible directions I could think of:
1. a real binary tree. `let level = leaves.map(create_node).collect(); while level.len() > 1 {level.chunks(2).map(create_internal_node).collect()}`
2. a vector of leaves and reply on recursive algorithms, probably plus a cache for performance (`HashMap<(HashValue, HashValue), HashValue)>`, for internal node hashing)
I guess it's easier saying than doing. So, your call. :P

phlip9 replied:
done! if you check out the MockTransactionAccumulator fixup! commit for the pure recursive implementation

msmouse replied:
That is elegant, overwhelmingly. Thanks ❤️.

xli commented:
Please add new response result type to json-rpc/types/proto/src/jsonrpc.proto

gregnazario commented on json-rpc/types/src/request.rs:
```diff
@@ -713,6 +728,64 @@ mod test {
     }
 
+    #[test]
+    fn get_accumulator_consistency_proof() {
+        // Array with all params
+        let value = serde_json::json!([11, 42]);
+        serde_json::from_value::<GetAccumulatorConsistencyProofParams>(value).unwrap();
+
+        // Array with too many params
+        let value = serde_json::json!([11, 42, 7]);
+        serde_json::from_value::<GetAccumulatorConsistencyProofParams>(value).unwrap_err();
```
Btw you could probably collapse a lot of this by making 2 vecs, 1 are successes, and the other are failures to iterate over.  Then just comment on the rows if you want.

bmwill replied:
If done naively that can end up making debugging errors more difficult since you then have a more difficult time seeing what the input is.

gregnazario commented on sdk/client/src/response.rs:
```diff
@@ -65,7 +65,11 @@ pub enum MethodResponse {
     GetCurrencies(Vec<CurrencyInfoView>),
     GetNetworkStatus(u64),
 
+    //
+    // Experimental APIs
+    //
```
No need to put this here, because if you add it, it probably won't be removed.  Additionally, what makes it experimental?  If I read this would I know if I could use it?

bmwill replied:
All the proof APIs are technically experimental (they're marked that way in the Jsonrpc docs) so they can have their apis tweaked at any point.

gregnazario commented on storage/diemdb/src/lib.rs:
```diff
@@ -764,17 +764,36 @@ impl DbReader for DiemDB {
+            // verify the latest LI nor verify a consistency proof up to the latest
+            // LI. If the client needs more epochs, we just return the consistency
+            // proof up to the last epoch change LI.
+            let verifiable_li = if epoch_change_proof.more {
```
Maybe I'm missing the context.  What's `more`?

phlip9 replied:
it's a signal in `EpochChangeProof` that the client needs to sync more epoch changes for them to fully catch up to the latest ledger state.

gregnazario commented on storage/diemdb/src/lib.rs:
```diff
@@ -764,17 +764,36 @@ impl DbReader for DiemDB {
+                    .ledger_info_with_sigs
+                    .last()
+                    .ok_or_else(|| format_err!(
+                        "No epoch changes despite claiming the client needs to sync more epochs: known_epoch={}, end_epoch={}",
```
Not just you, but I think we should probably audit / centralize errors / logs for consistency.

phlip9 replied:
definitely. IMO, diemdb should return concrete error types (vs. anyhow like it currently does), but that should probably come in a separate PR : )

bmwill commented on storage/diemdb/src/ledger_store/mod.rs:
```diff
@@ -310,10 +310,17 @@ impl LedgerStore {
     ) -> Result<AccumulatorConsistencyProof> {
-        Accumulator::get_consistency_proof(self, ledger_version + 1, client_known_version + 1)
+        let client_known_num_leaves = client_known_version
+            .map(|v| v.checked_add(1).ok_or_else(|| format_err!("overflow")))
```
IMO we should only be using "checked_X" rarely and try to prefer saturating variants. Of course this is situation dependent on what the code is trying to do

phlip9 replied:
either would work here IMO. calling any of the storage APIs with `u64::MAX - 1` is either a bug or someone trying to crash our node. saturating_add is easier to read at least

bmwill reviewed (commented):
PR is looking good and thanks for walking me through some of this PR yesterday, I don't have much more to add over what the other reviewers mentioned so lets get those comments addressed (plus fixing the conflict) and then we should be able to get this landed

kchalkias commented on sdk/client/src/verifying_client.rs:
```diff
@@ -105,42 +106,132 @@ impl<S: Storage> VerifyingClient<S> {
+        need_initial_accumulator: bool,
+    ) -> Result<(StateProof, Option<TransactionAccumulatorSummary>)> {
+        let (state_proof_view, state, maybe_consistency_proof_view) = if !need_initial_accumulator {
+            // just request the state proof, since we don't the initial accumulator
```
since we don't "need".

kchalkias commented on sdk/client/src/verifying_client.rs:
```diff
@@ -105,42 +106,132 @@ impl<S: Storage> VerifyingClient<S> {
+            .transpose()?;
+
+        // check the response metadata matches the state proof
+        verify_latest_li_matches_state(state_proof.0.ledger_info(), &state)?;
```
<3

kchalkias reviewed (commented):
I'm also happy crypto-wise! Great job

phlip9 commented:
rebased and addressed comments. note that fixup! [types] Add MockTransactionAccumulator makes large changes to the MockTransactionAccumulator; specifically, it now uses a purely recursive implementation as opposed to the previous Position-based implementation

msmouse commented on types/src/proof/definition.rs:
```diff
@@ -335,7 +337,10 @@ impl TransactionAccumulatorSummary {
+        let num_new_txns = target_li
+            .version()
+            .checked_sub(self.0.version())
+            .ok_or_else(|| format_err!("underflow"))?;
```
I think this is more of a "target_li older than my version" type of logic error that we want to be explicit.

phlip9 commented:
cc @bmwill Added the relaxed response ordering as we discussed : )

phlip9 commented:
Why does CI parse the doctests differently???!?!?!111?!1 hello???????

bmwill commented on storage/diemdb/src/ledger_store/mod.rs:
```diff
@@ -314,12 +314,9 @@ impl LedgerStore {
         let client_known_num_leaves = client_known_version
-            .map(|v| v.checked_add(1).ok_or_else(|| format_err!("overflow")))
-            .transpose()?
+            .map(|v| v.saturating_add(1))
```
I love saturating arithmetic so much more than checked arithmetic :)

bmwill commented on sdk/compatibility/src/lib.rs:
```diff
@@ -64,6 +64,7 @@ fn metadata_by_version() -> Result<()> {
 }
 
 #[test]
+#[ignore]
```
Once forge is finished we should be able to stop the need to ignore these

bmwill reviewed (approved):
Thanks for making the concurrency model a bit more relaxed, it should make usage more sane. I think this looks ready, assuming @msmouse is happy with the storage related changes.

msmouse commented:
Yup. Storage wise LGTM.

phlip9 commented:
/land
